libs/core/langchain_core/vectorstores/base.py

```diff
@@ -565,6 +516,54 @@ async def asearch(
                 "search_type to be 'similarity', 'similarity_score_threshold' or 'mmr'."
             )
 
+    def delete_by_filter(
+        self,
+        filter: Union[Dict[str, Any], List[Dict[str, Any]]],
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete documents by filter.
+
+        Args:
+            filter: Filter to apply to documents.
+            **kwargs: Other keyword arguments that subclasses might use.
+
+        Returns:
+            DeleteResponse: A response object that contains the list of IDs that were
+            successfully deleted from the vectorstore and the list of IDs that failed
+            to be deleted.
+
+        .. versionadded:: 0.2.15
+        """
+        raise NotImplementedError(
+            f"{self.__class__.__name__} does not yet support delete_by_filter."
+        )
+
+    def get_by_filter(
```

Comment:
Is there a way to filter and order by similarity? Would that be a `get_by_filter` with a `sort` based on the similarity? Or is that a similarity search with a filter? Something else?

The above makes me wonder if similarity search could be expressed as a sort, in which case basic similarity search is `filter = None, sort = Similarity`?

Reply:
I was thinking about introducing another method to capture search semantics. 


```python
def get_by_query(
        self,
        *,
        query: Q, # <-- e.g., text query or vector or BaseMedia (up to the implementation)
        filter: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = None,
        limit: Optional[int] = None,
        page: int = 0, # <-- Pagination added
        sort: Optional[Sort] = None,
        **kwargs,
) -> List[Hit]:  # <-- A hit would contain information like score, optional content etc.
```

----

i agree that sort and similarity can be viewed as the same thing. I need to think a bit more to see if we can merge them.

Reply:
@bjchambers Would standardization of filters, the ability to query by vector and the ability to retrieve vectors be sufficient to implement the search traversal patterns needed for GraphVectorStore?

At indexing time, the implementation can standardized fields into metadata; e.g., `namespace` (e.g., edge, node),  `edge_type` (just for edges), `edge_data` (e.g., the value of an href).

At query time, the implementation can rely on the ability to query by vector and filter on the standardized fields. Is that sufficient to implement GraphVectorStore? Is there a way to limit traversal by depth correctly (to avoid re-exploring nodes?)

Reply:
Also "query-by-ID" to get all the nodes with the selected IDs, but that could be part of the above. Currently, it does the traversal with just the IDs and then fetches those IDs. So maybe also:

1. `query-ids-by-vector`
2. `query-ids-by-filter`
3. `query-ids-by-vector-and-filter`
4. `query-by-ids`

The logic being that each node may be reached many times during the traversal -- rather than pulling back all the data each time, it is likely better to just pull back the data necessary for traversing. That is currently (i) the ID used for identifying the reachable nodes, (ii) the embedding (used for the MMR score calculation), and (iii) the outgoing tags (used for finding the next set of edges). So maybe a version that accepts a list of which fields to actually fetch (eg., we would only need the ID, the metadata["outgoing"], and the embedding). But that could perhaps be a future optimization...